lib/ansible/plugins/connection/ssh.py

```diff
@@ -505,7 +505,7 @@ def _run(self, cmd, in_data, sudoable=True):
             # and we've read all available output from it, we're done.
 
             if p.poll() is not None:
-                if not rpipes or not rfd:
+                if not rpipes or timeout == 0:
```

Comment:
This problem appears less frequently on Ansible 2.0 as the timeout value is much higher in select (ssh connection timeout + 2), but try with the following modification:
```python
            # Now we're awaiting_exit: has the child process exited? If it has,
            # and we've read all available output from it, we're done.
            time.sleep(.5)
            if p.poll() is not None:
                if not rpipes or not rfd:
                    break
```

Then execute the following playbook with ssh connection timeout set to 1:
```yaml
---
- hosts: all
  gather_facts: False
  tasks:
    - command: sleep {{ item | regex_replace('(\d)(\d)', '\1.\2') }}
      with_sequence: start=28 end=38
```

The output on my computer with Ansible 2.0.0.2:
```
$ ANSIBLE_TIMEOUT=1 ansible-playbook test.yml

PLAY [all] *********************************************************************

TASK [command] *****************************************************************
changed: [vagrant] => (item=28)
changed: [vagrant] => (item=29)
failed: [vagrant] (item=30) => {"failed": true, "item": "30", "module_stderr": "", "module_stdout": "", "msg": "MODULE FAILURE", "parsed": false}
failed: [vagrant] (item=31) => {"failed": true, "item": "31", "module_stderr": "", "module_stdout": "", "msg": "MODULE FAILURE", "parsed": false}
failed: [vagrant] (item=32) => {"failed": true, "item": "32", "module_stderr": "", "module_stdout": "", "msg": "MODULE FAILURE", "parsed": false}
failed: [vagrant] (item=33) => {"failed": true, "item": "33", "module_stderr": "", "module_stdout": "", "msg": "MODULE FAILURE", "parsed": false}
failed: [vagrant] (item=34) => {"failed": true, "item": "34", "module_stderr": "", "module_stdout": "", "msg": "MODULE FAILURE", "parsed": false}
changed: [vagrant] => (item=35)
changed: [vagrant] => (item=36)
changed: [vagrant] => (item=37)
changed: [vagrant] => (item=38)

NO MORE HOSTS LEFT *************************************************************
	to retry, use: --limit @test.retry

PLAY RECAP *********************************************************************
vagrant                    : ok=0    changed=0    unreachable=0    failed=1
```

Changing **not rfd** to **timeout == 0** guarantees that **p.stdout** is fully read and the error disappears.